project_analytic_code/models/project.py

```diff
@@ -0,0 +1,32 @@
+# Copyright 2021 Pierre Verkest
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models
+
+
+class Project(models.Model):
+    _inherit = "project.project"
+
+    analytic_account_code = fields.Char(
+        string="Analytic code",
+        related="analytic_account_id.code",
+        store=True,
+        index=True,
+    )
+
+    @api.model
+    def name_search(self, name="", args=None, operator="ilike", limit=100):
+        args = list(args or [])
+        if name:
+            args += ["|", ("analytic_account_code", "=", name)]
```

Comment:
```suggestion
            args = OR([args, [("analytic_account_code", "=", name)]]
```
This is not the best way to modify a domain, because depending on the provided one, it could lead to inconsistent results.
You should better use `OR` from `osv.expression` to safely "extend" the domain if such is your intent.

Reply:
hum... the matter is that I really want to add the OR operator with the domain add by super `("name", operator, name)`, I don't want to use OR with provide domain by args!... I'll revert this change !

Reply:
I'm not sure to understand why you don't want to use the `OR` operator, can you please explain?

The issue using `+` to extend the domain is that the ensuing domain will not necessarily be syntactically correct as the `|` expects two operands.

Please look at the following example done using odoo shell on a demo DB:
```
>>> domain = ['|', ('description', '!=', False), ('name', 'ilike', 'dummy')]
>>> self.env["project.project"].name_search(args=domain)
[]
>>> append_domain = domain + ['|', ('analytic_account_id.code', '=', '12')]
>>> append_domain
['|', ('description', '!=', False), ('name', 'ilike', 'dummy'), '|', ('analytic_account_id.code', '=', '12')]
>>> self.env["project.project"].name_search(args=append_domain)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/odoo/src/odoo/models.py", line 1902, in name_search
    ids = self._name_search(name, args, operator, limit=limit)
  File "/odoo/src/odoo/models.py", line 1918, in _name_search
    return self._search(args, limit=limit, access_rights_uid=name_get_uid)
  File "/odoo/src/odoo/models.py", line 4693, in _search
    if expression.is_false(self, args):
  File "/odoo/src/odoo/osv/expression.py", line 218, in is_false
    for token in reversed(normalize_domain(domain)):
  File "/odoo/src/odoo/osv/expression.py", line 210, in normalize_domain
    assert expected == 0, 'This domain is syntactically not correct: %s' % (domain)
AssertionError: This domain is syntactically not correct: ['|', ('description', '!=', False), ('name', 'ilike', 'dummy'), '|', ('analytic_account_id.code', '=', '12')]
```

OTOH, if you use the `OR` operator, you'll make sure of its correctness:
```
>>> from odoo.osv.expression import OR
>>> or_domain = OR([domain, [('analytic_account_id.code', '=', '12')]])
>>> or_domain
['|', '|', ('description', '!=', False), ('name', 'ilike', 'dummy'), ('analytic_account_id.code', '=', '12')
>>> self.env["project.project"].name_search(args=or_domain)
[]
```

Reply:
I agree that using osv.expression is safer but I'm not sure how I can do it in this case.

Domain I expect here is `["|", ("analytic_account_code", "=", name), ("name", "ilike", name)]` where the ` ("name", "ilike", name)` is provided by odoo https://github.com/OCA/OCB/blob/14.0/odoo/models.py#L1815

With your suggestion tests failed because assuming `args = []` it would produce the following domain at the end `[("analytic_account_code", "=", name), ("name", "ilike", name)]`.

In your odoo shell demonstration you haven't provide name parameter that why you get the error in this method if name is not provided this domain won't be added.

So I'm not sure how to express the OR with a part of the domain that I don't handle, any idea are welcome.

Thanks for your time

Reply:
Sorry for the delay.

I had a look in the core of Odoo and the good news, is that the possibility to override `name_search` was improved for v16.0 in this PR: https://github.com/odoo/odoo/pull/86588

Looking at it, we can see that there were already multiple overrides with the goal of searching on multiple fields and not only the `_rec_name`. Can we then mimic what is done in Odoo core to avoid the issue? Something around these lines:
```
    @api.model
    def _name_search(self, name, args=None, operator='ilike', limit=100, name_get_uid=None):
        if not (operator == 'ilike' and not (name or '').strip()):
            args = args or []
            domain = ['|', ('name', operator, name), ('analytic_account_code', operator, name)]
            return self._search(expression.OR([domain, args]), limit=limit, access_rights_uid=name_get_uid)
        return super()._name_search(name, args=args, operator=operator, limit=limit, name_get_uid=name_get_uid)
```
If it doesn't work as suggested, can you then please explain the issue in the roadmap segment of the readme so that we can merge this PR?

Reply:
sorry for the very, very late reply, your suggestion has been implemented from v16.0 and required change in base _name_search that I won't backport here!

As far I've seen practices in Odoo module is to not call super() which I don't find very kind in this case for this module while using _inherit.

I'm using it as it in production for years without any troubles. I'm on the way to migrate this module in version 17.0 so I'll take care of this.

Thanks for your time